File: OddEvenLinkedList/OddEvenLinkedList.cpp
/**
* Definition for singly-linked list.
*/
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
	ListNode* oddEvenList(ListNode* head) {
		if (head == nullptr || head->next == nullptr || head->next->next == nullptr)
			return head;
		int nodes_num = 0, num;
		ListNode* tail = head, *odd = head, *even = head->next;
		while (tail->next != nullptr) {
			nodes_num += 1;
			tail = tail->next;
		}
		if (nodes_num % 2 == 0)
			num = (nodes_num - 1) / 2;
		else
			num = nodes_num / 2;
		for (int i = 0; i <= num; i++) {
			odd->next = even->next;
			even->next = nullptr;
			tail->next = even;
			tail = tail->next;
			odd = odd->next;
			even = odd->next;
		}
		return head;
	}
};

File: OddEvenLinkedList/OddEvenLinkedList.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return head
        end = head
        len_list = 1
        while end.next is not None:
            end = end.next
            len_list += 1
        if len_list < 3:
            return head
        odd = head
        even = head.next
        count = 0
        if len_list % 2 == 0:
            num = len_list / 2
        else:
            num = (len_list - 1) / 2
        while count < num:
            odd.next = even.next
            even.next = None
            end.next = even
            end = end.next
            odd = odd.next
            even = odd.next
            count += 1
        return head